Sources/BashExpr/BashExpr.swift

```diff
@@ -0,0 +1,219 @@
+            return BashExprEvaluationResult(
+                stdout: out,
+                expr: .literal(out)
+            )
+        } else if commandName == "echo" {
+            let out = args.joined(separator: " ")
+            return BashExprEvaluationResult(
+                stdout: out,
+                expr: .literal(out)
+            )
+        } else {
+            return BashExprEvaluationResult(
+                expr: .error(
+                    BashExprError.configurationSpecificError(
+                        EvalError.notImplementedByConfiguration(commandName)
+                    )
+                ))
+        }
+    }
+}
+
+public struct CASToolEvalConfiguration: EvalConfiguration {
+    public enum EvalError: Equatable {
+        case sandboxDeny(commandName: String)
+    }
+
+    public let variableMappings: [String: BashExpr<EvalError>]
+
+    public init(variableMappings: [String: BashExpr<EvalError>]) {
+        self.variableMappings = variableMappings
+    }
+
+    public func runBuiltInCommand(commandName: String, args: [String]) async -> BashExprEvaluationResult<EvalError> {
+        if commandName == "castool" {
+            return BashExprEvaluationResult(expr: .literal("yay"))
+        } else if commandName == "echo" {
+            let out = args.joined(separator: " ")
+            return BashExprEvaluationResult(
+                stdout: out,
+                expr: .literal(out)
+            )
+        } else {
+            return BashExprEvaluationResult(
+                expr: .error(
+                    BashExprError.configurationSpecificError(
+                        EvalError.sandboxDeny(commandName: commandName)
+                    )
+                ))
+        }
+    }
+}
+
+public indirect enum BashExpr<T: Equatable>: Equatable {
+    case program([BashExpr])
+    case command(commandName: BashExpr, args: [BashExpr])
+    case literal(String)
+    case casRef(LLBDataID)
+    case number(Int)
+    case error(BashExprError<T>)
+    case concatenation([BashExpr])
+    case commandSubstitution([BashExpr])
+    case simpleExpansion([BashExpr])
+
+    public func eval<CE>(cfg: some EvalConfiguration<CE>) async -> BashExprEvaluationResult<CE>{
+        switch self {
+        case .program(let exprs):
+            guard exprs.count == 1, let expr = exprs.first else {
+                return .init(expr: .error(.notImplementedByEval("programs with more than one expr are not implemented")))
+            }
+            return await expr.eval(cfg: cfg)
+        case .number(let num):
+            return .init(expr: .literal("\(num)"))
+        case .simpleExpansion(let children):
+            guard case .literal("$") = children.first else {
+                return .init(
+                    expr: .error(
+                        .unexpectedToken(
+                            "\(children.first), expected .literal($)",
+                            NSRange(location: 0, length: 0)
```

Comment:
NSRange is a very strange type for Swift and should be avoided. `Range<Int>` sounds like the better choice here, right?

Reply:
Hmm.. Possibly. The location and length parameters are bogus too.